metricbeat/module/kvm/dommemstat/dommemstat.go

```diff
@@ -101,30 +103,40 @@ func (m *MetricSet) Fetch(report mb.ReporterV2) {
 
 		c, err = net.DialTimeout(u.Scheme, address, m.Timeout)
 		if err != nil {
-			report.Error(err)
+			report.Error(errors.Wrapf(err, "cannot connect to %v", u))
+			return
 		}
 	}
 
 	defer c.Close()
 
 	l := libvirt.New(c)
-	if err := l.Connect(); err != nil {
-		report.Error(err)
+	if err = l.Connect(); err != nil {
+		report.Error(errors.Wrap(err, "error connecting to libvirtd"))
+		return
 	}
+	defer func() {
+		if err = l.Disconnect(); err != nil {
+			report.Error(errors.Wrap(err, "failed to disconnect"))
+		}
+	}()
 
 	domains, err := l.Domains()
 	if err != nil {
-		report.Error(err)
+		report.Error(errors.Wrap(err, "error listing domains"))
+		return
 	}
 
 	for _, d := range domains {
 		gotDomainMemoryStats, err := l.DomainMemoryStats(d, maximumStats, flags)
 		if err != nil {
-			report.Error(err)
+			report.Error(errors.Wrapf(err, "error fetching memory stats for domain %s", d.Name))
+			continue
 		}
 
 		if len(gotDomainMemoryStats) == 0 {
-			report.Error(errors.New("no domain memory stats found"))
+			report.Error(fmt.Errorf("no memory stats for domain %s", d.Name))
```

Comment:
Personally I'd prefer errors.Errorf over fmt.Errorf in case we decide to consume the available stack information that it adds to the error.